pkg/providers/v2/instances.go

```diff
@@ -0,0 +1,365 @@
+		}
+
+		nextToken = response.NextToken
+		if aws.StringValue(nextToken) == "" {
+			break
+		}
+		request.NextToken = nextToken
+	}
+
+	if len(instances) == 0 {
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	if len(instances) > 1 {
+		return nil, fmt.Errorf("multiple instances found with private DNS name: %s", nodeName)
+	}
+
+	return instances[0], nil
+}
+
+// instanceShutdownByProviderID returns true if the instance is shutdown according to the cloud provider.
+func (i *instances) instanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error) {
+	ec2Instance, err := i.getInstanceByProviderID(ctx, providerID)
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			// We don't have live ec2Instance
+			return true, nil
+		}
+		return false, err
+	}
+
+	if ec2Instance.State != nil {
+		state := aws.StringValue(ec2Instance.State.Name)
+		if state == ec2.InstanceStateNameStopping || state == ec2.InstanceStateNameStopped {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// instanceShutdownByPrivateDNSName returns true if the instance is shutdown according to the cloud provider.
+func (i *instances) instanceShutdownByPrivateDNSName(ctx context.Context, nodeName string) (bool, error) {
+	ec2Instance, err := i.getInstanceByPrivateDNSName(ctx, nodeName)
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			// We don't have live ec2Instance
+			return true, nil
+		}
+		return false, err
+	}
+
+	if ec2Instance.State != nil {
+		state := aws.StringValue(ec2Instance.State.Name)
+		if state == ec2.InstanceStateNameStopping || state == ec2.InstanceStateNameStopped {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// nodeAddresses for Instance returns a list of v1.NodeAddress for the give instance.
+// TODO: should we support ExternalIP by default?
+func nodeAddressesForInstance(instance *ec2.Instance) ([]v1.NodeAddress, error) {
+	if instance == nil {
+		return nil, errors.New("provided instances is nil")
+	}
+
+	addresses := []v1.NodeAddress{}
+	for _, networkInterface := range instance.NetworkInterfaces {
+		// skip network interfaces that are not currently in use
+		if aws.StringValue(networkInterface.Status) != ec2.NetworkInterfaceStatusInUse {
+			continue
+		}
+
+		for _, privateIP := range networkInterface.PrivateIpAddresses {
+			if ipAddress := aws.StringValue(privateIP.PrivateIpAddress); ipAddress != "" {
+				ip := net.ParseIP(ipAddress)
+				if ip == nil {
+					return nil, fmt.Errorf("invalid IP address %q from instance %q", ipAddress, aws.StringValue(instance.InstanceId))
+				}
+
+				addresses = append(addresses, v1.NodeAddress{
+					Type:    v1.NodeInternalIP,
+					Address: ip.String(),
+				})
+			}
+		}
+	}
+
+	return addresses, nil
+}
+
+// getInstanceProviderID returns the provider ID of an instance which is ultimately set in the node.Spec.ProviderID field.
+// The well-known format for a node's providerID is:
+//    * aws:///<availability-zone>/<instance-id>
```

Comment:
@andrewsykim I checked the master node's providerID and it's of this form:
`aws:///<availability-zone>/<instance-id>`
instead of "aws://<availability-zone>/<instance-id>"
I'm wondering if I need to handle both cases or only the one with three "/"s